crates/matrix-sdk-crypto/src/identities/manager.rs

```diff
@@ -1019,6 +1040,126 @@ impl IdentityManager {
+    ///
+    /// Returns a list of modified InboundGroupSessions which need storing in
+    /// the database.
+    async fn update_sender_data_from_device_changes(
+        &self,
+        device_changes: &DeviceChanges,
+    ) -> Result<Vec<InboundGroupSession>, CryptoStoreError> {
+        let mut result = Vec::new();
+
+        for device in device_changes.new.iter().chain(device_changes.changed.iter()) {
+            // 1. Look for InboundGroupSessions from the device whose sender_data is
+            //    UnknownDevice. For such sessions, we now have the device, and can update
+            //    the sender_data accordingly.
+            //
+            // In theory, we only need to do this for new devices. In practice, I'm a bit
+            // worried about races leading us to getting stuck in the
+            // UnknownDevice state, so we'll paper over that by doing this check
+            // on device updates too.
+            let mut result1 = self
+                .update_sender_data_for_sessions_for_device(device, SenderDataType::UnknownDevice)
+                .await?;
+            result.append(&mut result1);
+
+            // 2. If, and only if, the device is now correctly cross-signed (ie,
+            //    device.is_cross_signed_by_owner() is true, and we have the master
+            //    cross-signing key for the owner), look for InboundGroupSessions from the
+            //    device whose sender_data is DeviceInfo. We can also update the sender_data
+            //    for these sessions.
+            //
+            // In theory, we can skip a couple of steps of the SenderDataFinder algorithm,
+            // because we're doing the cross-signing check here. In practice,
+            // it's *way* easier just to use the same logic.
+            let device_owner_identity = self.store.get_user_identity(device.user_id()).await?;
+            if device_owner_identity.is_some_and(|id| device.is_cross_signed_by_owner(&id)) {
+                let mut result1 = self
+                    .update_sender_data_for_sessions_for_device(device, SenderDataType::DeviceInfo)
+                    .await?;
+                result.append(&mut result1);
+            }
+        }
+
+        Ok(result)
+    }
+
+    /// Given a list of changed devices, look for [`InboundGroupSession`]s
+    /// whose sender data is in the given state, and update it.
+    ///
+    /// Returns a list of modified InboundGroupSessions which need storing in
+    /// the database.
+    #[instrument(skip(self))]
+    async fn update_sender_data_for_sessions_for_device(
+        &self,
+        device: &DeviceData,
+        sender_data_type: SenderDataType,
+    ) -> Result<Vec<InboundGroupSession>, CryptoStoreError> {
+        const IGS_BATCH_SIZE: usize = 50;
+
+        let mut result = Vec::new();
+        let Some(curve_key) = device.curve25519_key() else { return Ok(result) };
+
+        let mut last_session_id: Option<String> = None;
+        loop {
```

Comment:
This uses an unbounded amount of memory if there are lots of sessions for this device. Can we loop at a higher level to avoid this? We'll need to do the save_changes inside the loop I guess.

Reply:
Yeah. 74e63219d changes this so that we save each session separately. I'm not entirely convinced by that -- presumably there is a reason that we attempt to batch changes together into a single `save_changes`. But a middle ground is hard to achieve